src/vsr/replica.zig

```diff
@@ -2639,6 +3058,44 @@ pub fn Replica(
                 else => unreachable,
             }
 
+            // Don't participate in view change if the true op is not known and there is a fault to
+            // the right of the current op. The fault might be our true op, and sharing our current
+            // `replica.op` might cause the cluster's op to likewise regress.
+            //
+            // Note that for our purposes here, we only care about entries that were faulty after
+            // WAL recovery, not ones that were found to be faulty after the fact (e.g. due to
+            // `request_prepare`).
+            //
+            // Cases (`✓`: checkpoint, `✗`: faulty, `o`: `replica.op`):
+            // * ` ✓ o ✗ `: View change is unsafe.
+            // * ` ✗ ✓ o `: View change is unsafe.
+            // * ` ✓ ✗ o `: View change is safe.
+            // * ✓=o:       View change is unsafe.
+            //
+            // (But always allow the `start_view` — it will result in `op_known` being set to `true`).
+            if (!self.op_known and message.header.command != .start_view) {
+                assert(self.journal.recovered);
+                assert(self.journal.faulty.count > 0);
+
+                const end = self.journal.slot_for_op(self.op_checkpoint);
+                var slot = self.journal.slot_with_op(self.op).?;
+                while (true) {
```

Comment:
We should use a safety counter here, limited to how long we expect the loop to possibly run, in line with NASA's Power of Ten Rules, since we don't expect the loop to be infinite, i.e. it's not an event loop. There may be one or two while (true) loops in the code base still but it would be great to eliminate them.

Reply:
I rewrote this in https://github.com/coilhq/tigerbeetle/pull/91/commits/fda6cf6152a496abaec7f959f1678a6498de746d to use the bitset iterator.

Reply:
Really clean.